### ISSUE TITLE
How did you build this repo? 🙂

### ISSUE THREAD

**Issue by petre-c** [Filiprogrammer/SimpleJadePinServer#2]
@Filiprogrammer , I'm fascinated by the fact that you were able to build this repo.

Where did you learn about what was needed to build it? (I don't mean where did you learn to write code, but rather where did you learn about the workings of the pin server?)

I'd like to be able to do the same.. I'd be using C#, not python.

**Comment by Filiprogrammer** [comment]
I read the code of the original implementation of the pin server:
https://github.com/Blockstream/blind_pin_server

I also read the code for the client side on the Jade:
https://github.com/Blockstream/Jade/blob/master/main/process/pinclient.c

And it took a lot of trial and error over several weeks to get `SimpleJadePinServer` to work with my Jade.

> I'd like to be able to do the same.. I'd be using C#, not python.

This should be doable in C#, but you would have to find an alternative to [libwally](https://github.com/ElementsProject/libwally-core) for the cryptography, as it is only supported for C, C++, Python, Java & JavaScript.
Initially I started in C, but that was too cumbersome. So I started over in Python because it is relatively easy to get a proof of concept up and running.

**Comment by petre-c** [comment]
Understood, thank you.

What was your motivation though? Didn't the existing tools / methods suffice?

**Comment by Filiprogrammer** [comment]
> What was your motivation though? Didn't the existing tools / methods suffice?

Don't trust, verify ;)
I wanted to understand how it worked before I put money in my wallet. The most effective way to understand it was to reimplement it myself. With my own implementation and the understanding that the blind oracle is a completely safe alternative to a secure element, I now have the peace of mind that my wallet is secure.

And another important point was that the "official" web interface is just an [unmaintained branch](https://github.com/Blockstream/esp32_bc-ur/tree/emscripten) that is no longer up to date with [Blockstream's web UI](https://jadefw.blockstream.com/pinqr/index.html).

Also, `SimpleJadePinServer` has a very minimalistic code base, making it easier to audit and understand what is actually going on.

**Comment by petre-c** [comment]
Yeah, I had the same concerns, blind oracle was the reason I didn't use my Jade for many months when I first got it.

Ideally I wanted to do my signatures offline (on Tails + Electrum + Jade) and broadcast the signed transaction, which was impossible as blind oracle (non-QR-based) required access to the pinserver. 

As you mentioned, official QR-based unlock has a major risk - we don't know what information is being transmitted and we need to trust that blockstream won't 'steal' our money. 

Ultimately I created a multisig setup with official QR-unlock and a passphrase, which I enter on Jade after the unlock. Second signature happens offline and the risk of theft is low, I figured.

I'd still prefer to build my own QR-based blind oracle, just like yourself. 

Anyway, good job.
